Arm backend: Add INT16 support to rescale operation

Ninja91 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #13802
* #13801
* #13800
* #13799
* #13798
* #13797
* #13796

Add INT16 support for RequantizeNode rescale operations in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, cat, and FCNode operations, extending int16 support to RequantizeNode rescale operations.

Changes:
- Add INT16 dtype validation support in op_rescale.py
- Enable rescale operations for 16A8W quantization configuration

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency. RequantizeNode rescale operations are essential for proper quantization scaling in the 16A8W pipeline.

Differential Revision: [D80513725](https://our.internmc.facebook.com/intern/diff/D80513725/)

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218

Ninja91 committed:
Add INT16 support to rescale operation

Add INT16 support for RequantizeNode rescale operations in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, cat, and FCNode operations, extending int16 support to RequantizeNode rescale operations.

Changes:
- Add INT16 dtype validation support in op_rescale.py
- Enable rescale operations for 16A8W quantization configuration

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency. RequantizeNode rescale operations are essential for proper quantization scaling in the 16A8W pipeline.

Differential Revision: [D80513725](https://our.internmc.facebook.com/intern/diff/D80513725/)

[ghstack-poisoned]

digantdesai commented on backends/arm/operators/op_rescale.py:
```diff
@@ -46,13 +46,20 @@ def define_node(
         output_zp = cast(int, node.args[4])
 
-        if input_dtype != map_dtype(torch.int8, self.tosa_spec) and input_zp != 0:
+        if (
```
I think 8 -> 16 and 16 -> 8 both are supported for the INT profile (and INT16)? Can you make sure you check?

digantdesai reviewed (changes requested):
I didn't see torch ops like tests for TOSA dialect ops, perhaps we should add it here. @per what do you think?

Ninja91 committed:
Update on "Arm backend: Add INT16 support to rescale operation"


Add INT16 support for RequantizeNode rescale operations in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, cat, and FCNode operations, extending int16 support to RequantizeNode rescale operations.

Changes:
- Add INT16 dtype validation support in op_rescale.py
- Enable rescale operations for 16A8W quantization configuration

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency. RequantizeNode rescale operations are essential for proper quantization scaling in the 16A8W pipeline.

Differential Revision: [D80513725](https://our.internmc.facebook.com/intern/diff/D80513725/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

digantdesai reviewed (approved):
Review automatically exported from Phabricator review in Meta.

Ninja91 committed:
Update on "Arm backend: Add INT16 support to rescale operation"


Add INT16 support for RequantizeNode rescale operations in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, cat, and FCNode operations, extending int16 support to RequantizeNode rescale operations.

Changes:
- Add INT16 dtype validation support in op_rescale.py
- Enable rescale operations for 16A8W quantization configuration

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency. RequantizeNode rescale operations are essential for proper quantization scaling in the 16A8W pipeline.

Differential Revision: [D80513725](https://our.internmc.facebook.com/intern/diff/D80513725/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add INT16 support to rescale operation"


Add INT16 support for RequantizeNode rescale operations in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, cat, and FCNode operations, extending int16 support to RequantizeNode rescale operations.

Changes:
- Add INT16 dtype validation support in op_rescale.py
- Enable rescale operations for 16A8W quantization configuration

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency. RequantizeNode rescale operations are essential for proper quantization scaling in the 16A8W pipeline.

Differential Revision: [D80513725](https://our.internmc.facebook.com/intern/diff/D80513725/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add INT16 support to rescale operation"


Add INT16 support for RequantizeNode rescale operations in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, cat, and FCNode operations, extending int16 support to RequantizeNode rescale operations.

Changes:
- Add INT16 dtype validation support in op_rescale.py
- Enable rescale operations for 16A8W quantization configuration

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency. RequantizeNode rescale operations are essential for proper quantization scaling in the 16A8W pipeline.

Differential Revision: [D80513725](https://our.internmc.facebook.com/intern/diff/D80513725/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add INT16 support to rescale operation"


Add INT16 support for RequantizeNode rescale operations in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, cat, and FCNode operations, extending int16 support to RequantizeNode rescale operations.

Changes:
- Add INT16 dtype validation support in op_rescale.py
- Enable rescale operations for 16A8W quantization configuration

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency. RequantizeNode rescale operations are essential for proper quantization scaling in the 16A8W pipeline.

Differential Revision: [D80513725](https://our.internmc.facebook.com/intern/diff/D80513725/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add INT16 support to rescale operation"


Add INT16 support for RequantizeNode rescale operations in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, cat, and FCNode operations, extending int16 support to RequantizeNode rescale operations.

Changes:
- Add INT16 dtype validation support in op_rescale.py
- Enable rescale operations for 16A8W quantization configuration

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency. RequantizeNode rescale operations are essential for proper quantization scaling in the 16A8W pipeline.

Differential Revision: [D80513725](https://our.internmc.facebook.com/intern/diff/D80513725/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]

Ninja91 committed:
Update on "Arm backend: Add INT16 support to rescale operation"


Add INT16 support for RequantizeNode rescale operations in ExecutorTorch ARM backend.

This follows the pattern established for linear, mul, sigmoid, tanh, slice, view/transpose, cat, and FCNode operations, extending int16 support to RequantizeNode rescale operations.

Changes:
- Add INT16 dtype validation support in op_rescale.py
- Enable rescale operations for 16A8W quantization configuration

The 16A8W configuration uses 16-bit activations with 8-bit weights, enabling higher precision for activations while maintaining weight efficiency. RequantizeNode rescale operations are essential for proper quantization scaling in the 16A8W pipeline.

Differential Revision: [D80513725](https://our.internmc.facebook.com/intern/diff/D80513725/)

cc digantdesai freddan80 per zingo oscarandersson8218

[ghstack-poisoned]
